In [22]:
import numpy as np
import torch

## Tensors basics

In [2]:
x = torch.empty(2, 5)

In [3]:
x

tensor([[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]])

In [4]:
zeros = torch.zeros(2, 3)

In [5]:
ones = torch.ones(2, 3)

In [6]:
torch.manual_seed(42)

In [7]:
rnd = torch.rand(2, 3)

In [8]:
zeros, ones, rnd

(tensor([[0., 0., 0.],
         [0., 0., 0.]]),
 tensor([[1., 1., 1.],
         [1., 1., 1.]]),
 tensor([[0.8823, 0.9150, 0.3829],
         [0.9593, 0.3904, 0.6009]]))

In [9]:
x = torch.empty(2, 2, 3)

In [10]:
x.size()

torch.Size([2, 2, 3])

In [11]:
x.shape

torch.Size([2, 2, 3])

In [12]:
empty_like_x = torch.empty_like(x)

In [13]:
zeros_like_x = torch.zeros_like(x)

In [14]:
ones_like_x = torch.ones_like(x)

In [15]:
rand_like_x = torch.rand_like(x)

In [16]:
empty_like_x, zeros_like_x, ones_like_x, rand_like_x

(tensor([[[0., 0., 0.],
          [0., 0., 0.]],
 
         [[0., 0., 0.],
          [0., 0., 0.]]]),
 tensor([[[0., 0., 0.],
          [0., 0., 0.]],
 
         [[0., 0., 0.],
          [0., 0., 0.]]]),
 tensor([[[1., 1., 1.],
          [1., 1., 1.]],
 
         [[1., 1., 1.],
          [1., 1., 1.]]]),
 tensor([[[0.2566, 0.7936, 0.9408],
          [0.1332, 0.9346, 0.5936]],
 
         [[0.8694, 0.5677, 0.7411],
          [0.4294, 0.8854, 0.5739]]]))

In [19]:
some_constants = torch.tensor([[3.1415926, 2.71828], [1.61803, 0.0072897]])

In [17]:
some_integers = torch.tensor((2, 3, 5, 7, 11, 13, 17, 19))

In [18]:
more_integers = torch.tensor(((2, 4, 6), [3, 6, 9]))

In [20]:
some_constants, some_integers, more_integers

(tensor([[3.1416, 2.7183],
         [1.6180, 0.0073]]),
 tensor([ 2,  3,  5,  7, 11, 13, 17, 19]),
 tensor([[2, 4, 6],
         [3, 6, 9]]))

In [23]:
numpy_array = np.ones((2, 3))

In [24]:
pytorch_tensor = torch.from_numpy(numpy_array)

In [25]:
pytorch_tensor

tensor([[1., 1., 1.],
        [1., 1., 1.]], dtype=torch.float64)

In [26]:
pytorch_tensor.numpy()

array([[1., 1., 1.],
       [1., 1., 1.]])

In [27]:
type(pytorch_tensor.numpy())

numpy.ndarray

Tensor data types

In [28]:
a = torch.ones((2, 3), dtype=torch.int16)

In [29]:
a.dtype

torch.int16

In [31]:
b = torch.rand((2, 3), dtype=torch.float64) * 20

In [32]:
c = b.to(torch.int32)

In [33]:
b, c

(tensor([[12.2487,  1.7622, 14.0236],
         [12.4681,  8.7456,  1.4947]], dtype=torch.float64),
 tensor([[12,  1, 14],
         [12,  8,  1]], dtype=torch.int32))

In [35]:
torch.cuda.is_available()

False

In [38]:
b.to("cpu")

tensor([[12.2487,  1.7622, 14.0236],
        [12.4681,  8.7456,  1.4947]], dtype=torch.float64)

Available data types include:

* `torch.bool`

torch.int8

torch.uint8

torch.int16

torch.int32

torch.int64

torch.half

torch.float

torch.double

torch.bfloat

## Math & Logic with Tensors

In [39]:
ones = torch.zeros(2, 2) + 1

In [40]:
twos = torch.ones(2, 2) * 2

In [41]:
threes = (torch.ones(2, 2) * 7 - 1) / 2

In [42]:
fours = twos ** 2

In [43]:
sqrt2s = twos ** 0.5

In [44]:
powers2 = twos ** torch.tensor([[1, 2], [3, 4]])

In [45]:
fives = ones + fours

In [46]:
dozens = threes * fours

In [47]:
ones, twos, threes, fours

(tensor([[1., 1.],
         [1., 1.]]),
 tensor([[2., 2.],
         [2., 2.]]),
 tensor([[3., 3.],
         [3., 3.]]),
 tensor([[4., 4.],
         [4., 4.]]))

In [48]:
sqrt2s, powers2, fives, dozens

(tensor([[1.4142, 1.4142],
         [1.4142, 1.4142]]),
 tensor([[ 2.,  4.],
         [ 8., 16.]]),
 tensor([[5., 5.],
         [5., 5.]]),
 tensor([[12., 12.],
         [12., 12.]]))

In [49]:
x = torch.empty(2, 3)

In [57]:
x.fill_(1.125)

tensor([[1.1250, 1.1250, 1.1250],
        [1.1250, 1.1250, 1.1250]])

In [58]:
x

tensor([[1.1250, 1.1250, 1.1250],
        [1.1250, 1.1250, 1.1250]])

In [59]:
x.mean()

tensor(1.1250)

In [51]:
x.std()

tensor(0.)

In [60]:
x.sum()

tensor(6.7500)

In [53]:
x.sum().item()

0.0

In [54]:
type(x.sum().item())

float

In [66]:
x = torch.tensor([[11, 12, 13], [21, 22, 23]], dtype=torch.uint8)

In [62]:
x

tensor([[11, 12, 13],
        [21, 22, 23]])

In [63]:
x[1, 2]

tensor(23)

In [64]:
x[:, 2]

tensor([13, 23])

In [67]:
x.dtype

torch.uint8

In [68]:
x.device

device(type='cpu')

## Component-wise and vector/matrix operations

In [69]:
x = torch.tensor([ 10., 20., 30.])

In [70]:
y = torch.tensor([ 11., 21., 31.])

In [71]:
x + y

tensor([21., 41., 61.])

In [72]:
x * y

tensor([110., 420., 930.])

In [73]:
x ** 2

tensor([100., 400., 900.])

In [74]:
m = torch.tensor([[ 0., 0., 3. ],
                  [ 0., 2., 0. ],
                  [ 1., 0., 0. ]])

In [75]:
m.mv(x)

tensor([90., 40., 10.])

In [76]:
m @ x

tensor([90., 40., 10.])

The @ operator corresponds to matrix/vector or matrix/matrix multiplication, while * is component-wise product and can be applied to tensors of arbitrary size, in particular of dimension greater than 2.

## Standard linear operations

In [77]:
y = torch.randn(3)

In [78]:
y

tensor([-0.3514, -0.7906, -0.0915])

In [79]:
m = torch.rand(3, 3)

In [80]:
m

tensor([[0.0050, 0.3068, 0.1165],
        [0.9103, 0.6440, 0.7071],
        [0.6581, 0.4913, 0.8913]])

In [81]:
q = torch.linalg.lstsq(m, y).solution

In [82]:
m @ q

tensor([-0.3514, -0.7906, -0.0915])

In [83]:
q

tensor([-0.7659, -1.6549,  1.3751])

## Transpose, view, indexing

In [84]:
output3d = torch.rand(6, 20, 20)

In [85]:
input1d = output3d.reshape(6 * 20 * 20)

In [86]:
torch.reshape(output3d, (6 * 20 * 20,)).shape

torch.Size([2400])

In [88]:
output3d.view(6 * 20 * 20).shape

torch.Size([2400])

In [89]:
x = torch.tensor([ [ 1, 3, 0 ],
                   [ 2, 4, 6 ] ])

In [90]:
x.t()

tensor([[1, 2],
        [3, 4],
        [0, 6]])

In [91]:
x.view(-1)

tensor([1, 3, 0, 2, 4, 6])

In [92]:
x.view(3, -1)

tensor([[1, 3],
        [0, 2],
        [4, 6]])

In [93]:
x[:, 1:3]

tensor([[3, 0],
        [4, 6]])

In [94]:
x.view(1, 2, 3).expand(3, 2, 3)

tensor([[[1, 3, 0],
         [2, 4, 6]],

        [[1, 3, 0],
         [2, 4, 6]],

        [[1, 3, 0],
         [2, 4, 6]]])

In [95]:
x.unsqueeze(0).expand(3, 2, 3)

tensor([[[1, 3, 0],
         [2, 4, 6]],

        [[1, 3, 0],
         [2, 4, 6]],

        [[1, 3, 0],
         [2, 4, 6]]])

In [96]:
x = torch.tensor([ [ [ 1, 2, 1 ],
                     [ 2, 1, 2 ] ],
                   [ [ 3, 0, 3 ],
                     [ 0, 3, 0 ] ] ])

In [97]:
x[0:1, :, :]

tensor([[[1, 2, 1],
         [2, 1, 2]]])

In [98]:
x[:, :, 0:2]

tensor([[[1, 2],
         [2, 1]],

        [[3, 0],
         [0, 3]]])

In [99]:
x.transpose(0, 1)

tensor([[[1, 2, 1],
         [3, 0, 3]],

        [[2, 1, 2],
         [0, 3, 0]]])

In [100]:
x.transpose(0, 2)

tensor([[[1, 3],
         [2, 0]],

        [[2, 0],
         [1, 3]],

        [[1, 3],
         [2, 0]]])

In [101]:
x.transpose(1, 2)

tensor([[[1, 2],
         [2, 1],
         [1, 2]],

        [[3, 0],
         [0, 3],
         [3, 0]]])

For efficiency reasons, different tensors can share the same data and modifying one will modify the others. By default do not make the assumption that two tensors refer to different data in memory.

In [102]:
a = torch.full((2, 3), 1)

In [103]:
a

tensor([[1, 1, 1],
        [1, 1, 1]])

In [104]:
b = a.view(-1)

In [105]:
b

tensor([1, 1, 1, 1, 1, 1])

In [106]:
a[1, 1] = 2

In [107]:
a

tensor([[1, 1, 1],
        [1, 2, 1]])

In [108]:
b

tensor([1, 1, 1, 1, 2, 1])

In [109]:
b[0] = 9

In [111]:
a

tensor([[9, 1, 1],
        [1, 2, 1]])

In [112]:
b

tensor([9, 1, 1, 1, 2, 1])

## Einstein summation: matrix product

In [113]:
p = torch.rand(2, 5)

In [114]:
q = torch.rand(5, 4)

In [115]:
torch.einsum('ij,jk->ik', p, q)

tensor([[1.4122, 1.3369, 1.1461, 2.0365],
        [1.1648, 1.1237, 0.9161, 1.5248]])

In [116]:
p@q

tensor([[1.4122, 1.3369, 1.1461, 2.0365],
        [1.1648, 1.1237, 0.9161, 1.5248]])

## Tensors internals

In [117]:
x = torch.zeros(2, 4)

In [118]:
x.storage()

/var/folders/8p/sj6ncwq12_z6dgkmp_9c7llc0000gn/T/ipykernel_94488/2674773268.py:1: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  x.storage()


 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
[torch.storage.TypedStorage(dtype=torch.float32, device=cpu) of size 8]

In [119]:
q = x.storage()

In [120]:
q[4] = 1.0

In [121]:
x

tensor([[0., 0., 0., 0.],
        [1., 0., 0., 0.]])

In [122]:
q = torch.arange(0., 20.).storage()

In [123]:
x = torch.empty(0).set_(q, storage_offset = 5, size = (3, 2), stride = (4, 1))

In [125]:
x

tensor([[ 5.,  6.],
        [ 9., 10.],
        [13., 14.]])

In [126]:
n = torch.linspace(1, 4, 4)

In [127]:
n

tensor([1., 2., 3., 4.])

In [128]:
torch.tensor(0.).set_(n.storage(), 1, (3, 3), (0, 1))

tensor([[2., 3., 4.],
        [2., 3., 4.],
        [2., 3., 4.]])

In [129]:
torch.tensor(0.).set_(n.storage(), 1, (2, 4), (1, 0))

tensor([[2., 2., 2., 2.],
        [3., 3., 3., 3.]])

## Torchvision datasets

In [130]:
import torchvision

In [131]:
cifar = torchvision.datasets.CIFAR10('./cifar10/', train = True, download = True)

Files already downloaded and verified


In [135]:
cifar.classes

['airplane',
 'automobile',
 'bird',
 'cat',
 'deer',
 'dog',
 'frog',
 'horse',
 'ship',
 'truck']

In [137]:
cifar.data[10].shape

(32, 32, 3)

In [138]:
cifar.data[10, 15, 15]

array([74, 69, 50], dtype=uint8)

In [148]:
x = torch.from_numpy(cifar.data).permute(0, 3, 1, 2).float() / 255

Putting `float() / 255` in the end of a line normalizes images.

In [140]:
x.dtype, x.size(), x.min().item(), x.max().item()

(torch.float32, torch.Size([50000, 3, 32, 32]), 0.0, 1.0)

In [141]:
cifar.data[0, 5, 9]

array([122,  82,  44], dtype=uint8)

Note that there are different storage conventions between some libraries used by PyTorch (pillow and NumPy) and PyTorch itself:
* loading the images yields a tensor of shape 50000×32×32×3, but
* PyTorch works with the channel dimension as the second one: 50000×3×32×32.

This change is made with `permute(0, 3, 1, 2)` which means that we want dimension 3 of the original tensor to lie at the second position of the new tensor.

In the original tensor, accessing pixel `(5, 9)` of the first image `cifar.data[0, 5, 9]` returns
`[122 82 44]`, because the last dimension is the number of channels.

Once the permutation is done, `x[n, 0, :, :]` allows to access the red channel of image `n`.

Narrows to the first images, converts to float

In [142]:
x = x[:48]

Saves these samples as a single image

In [143]:
torchvision.utils.save_image(x, 'cifar-4x12.png',
                             nrow = 12, pad_value = 1.0)

Switches the row and column indexes

In [144]:
x.transpose_(2, 3)

tensor([[[[0.2314, 0.0627, 0.0980,  ..., 0.8157, 0.7059, 0.6941],
          [0.1686, 0.0000, 0.0627,  ..., 0.7882, 0.6784, 0.6588],
          [0.1961, 0.0706, 0.1922,  ..., 0.7765, 0.7294, 0.7020],
          ...,
          [0.6196, 0.4824, 0.4627,  ..., 0.6275, 0.7216, 0.8471],
          [0.5961, 0.4667, 0.4706,  ..., 0.2196, 0.3804, 0.5922],
          [0.5804, 0.4784, 0.4275,  ..., 0.2078, 0.3255, 0.4824]],

         [[0.2431, 0.0784, 0.0941,  ..., 0.6667, 0.5451, 0.5647],
          [0.1804, 0.0000, 0.0275,  ..., 0.6000, 0.4824, 0.5059],
          [0.1882, 0.0314, 0.1059,  ..., 0.6314, 0.5647, 0.5569],
          ...,
          [0.5176, 0.3451, 0.3294,  ..., 0.5216, 0.5804, 0.7216],
          [0.4902, 0.3255, 0.3294,  ..., 0.1216, 0.2431, 0.4627],
          [0.4863, 0.3412, 0.2863,  ..., 0.1333, 0.2078, 0.3608]],

         [[0.2471, 0.0784, 0.0824,  ..., 0.3765, 0.3765, 0.4549],
          [0.1765, 0.0000, 0.0000,  ..., 0.1333, 0.1647, 0.3686],
          [0.1686, 0.0000, 0.0314,  ..., 0

In [145]:
torchvision.utils.save_image(x, 'cifar-4x12-rotated.png',
                                        nrow = 12, pad_value = 1.0)

Since the data follows the standard PyTorch “channel first” convention, transposing dimensions 2 and 3 (that is the 3rd and the fourth) exchanges the height and width of the images.

Remember that methods ending with an underscore operate in-place.

Kills the green and blue channels

In [146]:
x[:, 1:3].fill_(0)

tensor([[[[0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          ...,
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.]],

         [[0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          ...,
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.]]],


        [[[0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          ...,
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.]],

         [[0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          ...,
          [0., 0., 0.,  ..., 0., 0., 0.],
        

In [147]:
torchvision.utils.save_image(x, 'cifar-4x12-rotated-and-red.png',
                                        nrow = 12, pad_value = 1.0)

Here, we set all the values of the green and blue channels to zero (channels 1 and 2 respectively).

In [149]:
x = torch.from_numpy(cifar.data).permute(0, 3, 1, 2).float() / 255

In [150]:
x = x[:48]

In [151]:
x[:, 1:3, 0:20, 0:20].fill_(0)

tensor([[[[0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          ...,
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.]],

         [[0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          ...,
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.]]],


        [[[0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          ...,
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.]],

         [[0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          [0., 0., 0.,  ..., 0., 0., 0.],
          ...,
          [0., 0., 0.,  ..., 0., 0., 0.],
        

In [152]:
torchvision.utils.save_image(x, 'cifar-4x12-partially-red.png', nrow = 12, pad_value = 1.0)

## Useful links

1. [PyTorch deeper tutorial on tensors](https://pytorch.org/tutorials/beginner/introyt/tensors_deeper_tutorial.html)
2. kdnuggets: [PyTorch tensors basics](https://www.kdnuggets.com/2018/05/pytorch-tensor-basics.html)
3. Medium: [Tensors basics](https://medium.com/codex/tensor-basics-in-pytorch-252a34288f2)